packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,274 @@
+                        };
+                        this.fireTaskCreatedEvent(newTask);
+                        resolve(newTask);
+
+                    } catch (error) {
+                        this.logger.error(`Error occured while creating task: ${error}`);
+                        reject(new Error(error));
+                    }
+                })
+                .catch(err => {
+                    reject(new Error(err));
+                });
+
+        });
+    }
+
+    protected fireTaskExitedEvent(event: ITaskExitedEvent) {
+        this.logger.debug(log => log(`task has exited:`, event));
+        // notify all clients. They can filter on context to see if it's relevant to them
+        this.clients.forEach(client => {
+            client.onTaskExit(event);
+        });
+    }
+
+    protected fireTaskCreatedEvent(event: ITaskInfo) {
+        this.logger.debug(log => log(`task created:`, event));
+        // notify all clients. They can filter on context to see if it's relevant to them
+        this.clients.forEach(client => {
+            client.onTaskCreated(event);
+        });
+    }
+
+    kill(id: number): Promise<void> {
+        const taskToKill = this.taskManager.get(id);
+        if (taskToKill !== undefined) {
+            taskToKill.kill();
+            this.logger.info(`Attempting to kill task id ${id}`);
+        }
+        return Promise.resolve();
+    }
+
+    /** Set the client we send the notifications-to. We expect that there might be
+     * multiple clients, so add them to a list as they are set
+     */
+    setClient(client: ITaskClient) {
+        this.clients.push(client);
+    }
+
+    /**
+     * uses heuristics to look-for a command. Will look into the system path, if the command
+     * is given without a path. Will resolve if a potential match is found, else reject. There
+     * is no garantee that a command we find will be the one executed, if multiple commands with
+     * the same name exist.
+     * @param command command name to look-for
+     * @param cwd current working directory
+     */
+    private async findCommand(command: string, cwd: string): Promise<string> {
+        const systemPath = process.env.PATH;
+
+        return new Promise<string>(async (resolve, reject) => {
+
+            if (path.isAbsolute(command)) {
+                if (await this.fileExists(command)) {
+                    resolve(command);
+                    return;
+                } else {
+                    const uri = new URI(cwd).resolve(command);
+                    const resolved_command = FileUri.fsPath(uri);
+                    if (await this.fileExists(resolved_command)) {
+                        resolve(resolved_command);
+                        return;
+                    }
+                }
+            } else {
+                // look for command relative to cwd
+                const resolved_command = FileUri.fsPath(new URI(cwd).resolve(command));
```

Comment:
About CWD should it default to workspaceRoot somewhere? I don't think it makes sense to use the process.cwd

Reply:
With the refactorings to support multiple clients, the Task backend server has become stateless, with regards to the workspace being used (it used to be that each client would get it's own instance of the Task server, that would know what the WS Root was). The backend doesn't know which workspace(s) are being used. 

Instead, each task configuration defines what "CWD" should be used by that task. The F-E will replace the symbolic value "$workspace" by its value, if found in the task configuration's cwd. In the end, that same CWD value is used to create the process, used by the task.

Reply:
update: now cwd defaults to the WS Root, unless otherwise defined in the task configuration (set by frontend).